### DISCUSSION TITLE
One of the constraints evaluates to a FALSE LOGICAL variable. Your model is infeasible

### DISCUSSION THREAD

**Discussion by SIYUANLI2023** [yalmip/YALMIP#1385] [Optimization modelling]
Dear Johan,

I apologize for any inconvenience, but I am seeking assistance with an issue I've encountered while using the YALMIP toolbox for solving a Mixed Integer Linear Programming (MILP) problem. My problem involves several constraints, as illustrated in Figure 1. Additionally, the definitions for x{k}, mdps(k), and Pflag_trafficlights are provided as follows, respectively. Here,  S1=[1,0,0], and Pflag_trafficlights can be known in advance.

x = sdpvar(repmat(ns,1,H+1),ones(1,H+1)); 
mdps=intvar(H+1,1); 
if strcmp(traffic_light_statuses, 'GREEN')
    Pflag_trifficlights(k)=1;
elseif strcmp(traffic_light_statuses, 'YELLOW')
    Pflag_trifficlights(k)=2;
else
        Pflag_trifficlights(k)=3;
end

Unfortunately, I am struggling to identify the cause of an error in my implementation and am unsure how to proceed. Is there a mistake in the way my constraints are described? Is it possible that the cause of this error is caused by a conflict with another constraint. Could you please offer some guidance on how to address this issue? I would greatly appreciate any advice you can provide. Thank you very much!

Warm regards,
Lee.

![image](https://github.com/yalmip/YALMIP/assets/148333150/12e2e0a0-466b-4a8c-8330-4fd87b1d5f5d)

**Comment by johanlofberg** [comment]
https://yalmip.github.io/trivialfalse

**Reply by SIYUANLI2023** [reply]
Hi Johan, thanks for your reply. Do you think the error is due to the fact that S1*x{k} may not be in (0,1)? I would like to describe that mdps=2 when Pflag_trifficlights(k)==2 || Pflag_trifficlights(k)==3 and -1<=S1*x{k}<=0. Can I ask how the correct one should be described, I am sorry for that, I just started to work on this area, I can be unskilled in some places, thanks for your for your help and reply.

![image](https://github.com/yalmip/YALMIP/assets/148333150/a2b2f054-0af0-4253-8db9-e96e55df0a78)

**Reply by johanlofberg** [reply]
It has nothing to do about any other constraints or integrality or anything. The expression you have added to the model does not contain any decision variables but has evaluated to a constant value such as 5 * x <= 5 * x-3 which thus immediately evaluates to the matlab code 0<=-3 which makes no sense to try to add to a model. For that to happen x{k} must contain some constant data which turns S1 * x{k} to a constant which doesn't satisfy the constraints -1<=S1 * x <=0. You will simply have to look at what the expression is when it fails

**Reply by SIYUANLI2023** [reply]
Hi Johan, thanks for your patient reply, but I still don't quite get it, in the code I got x = sdpvar(repmat(ns,1,H+1),ones(1,H+1)); doesn't that mean that x is the decision variable? This is an MPC optimization problem with k in(1,N),N is the prediction time domain, and each time it enters this optimization problem, only x(1) is known, x(2).... .x(N) are all unknown and all must be determined from mdp(k). 

But your mention of -1<=S1*x{k}<=0 not being satisfied does exist, and at the initial moment such a condition is not satisfied, but S1*x{k} is updated with the event into (-1, 0), and the logic I'm trying to describe when it is only possible if (Pflag_trifficlights(k)==2 || Pflag_trifficlights(k )==3) and (-1<=S1*x{k}<=0) hold at the same time, mdp(k)==2,Is there something wrong with my code in describing the logic?

The strange thing is that the code is running normally in the beginning, and only after some time, this error is reported, which means that this code has been executed many times, that's why I'm guessing that it's not an error due to a conflict with other constraints.

**Reply by johanlofberg** [reply]
hence have you run with your code with printouts of the stuff that you have involved in the constraints, so that you see what they are when the error occurs, because they are not what you think

other constraints has nothing to do with it, because the error is that you try to add stuff which simply doesn't make sense. YALMIP complains like this when you try to write things like [constraints, 1 <= 0]

**Reply by SIYUANLI2023** [reply]
Hi Johan, thanks a lot! I printed the result and tried it a few times and found that the code stops every time S1*x{k} gets close to -2, which doesn't make sense, it belongs to (-5,0). This makes me unsure of the next step to debug my code, can you give me some further help? What is your opinion on this situation?
![image](https://github.com/yalmip/YALMIP/assets/148333150/5b414157-edc9-4573-b7cc-0230e5cd864f)

**Reply by johanlofberg** [reply]
Well clearly then it makes absolutely no sense that you are trying to impose a constraint on a constant. It makes absolutely no sense to say that you want [-1 <= -2.1845 <=0]. A number is what it is, and you cannot change that fact, and it makes no sense that it should be part of a set of constraints. So to make it clear: you have probably accidentally inserted a constant into the previous variable x{k} somewhere in your code

**Reply by SIYUANLI2023** [reply]
Hi Johan. thanks for the reply, I'll try to go ahead and make the question a little clearer.

First of all, S1*x{k} represents the position of the car, and this value -2.1845 actually represents the initial position, which is S1*x{1}, and this initial position is related to the underlying control, which can be obtained in real time by some means. Of course, this initial position is not the absolute meaning of the initial position, every time I execute this program, the real-time position of the cart as the initial point, which is why S1*x{k} is constantly updated. Inside my program, I need to make a real-time prediction of the subsequent position, and this prediction model is not deterministic and related to mdp(k), which means that the value of the subsequent S1*x{k} is also not deterministic and constant, which is why x{k} is designed to be the decision variable.  

-2.1845 is actually the real-time position of the cart at moment t, in the interval (-5, 0). What I want to do is that if Pflag_trifficlights(k)==2 || Pflag_trifficlights(k)==3 is satisfied and S1*x{1} belongs to (-5, 0), then the value of mdp(k) becomes 2. Now Pflag_trifficlights(k)==2 || Pflag_ trifficlights(k)==3 is satisfied, and S1*x{1} belongs to (-5, 0) is also satisfied, but the program reports an error, which puzzles me.

**Reply by johanlofberg** [reply]
You will have to encode the initial value of the car by adding a constraint x{1} == initialvalue, instead of changing its value x{1} = initialvalue, since you will not be able to append a trivially false constant logical expression to a model

**Reply by SIYUANLI2023** [reply]
Hi, Johan! Thank you! I changed like you said, constraints = [constraints,x{1}==[S1*Rx;S2*Rx;S4*Rx];];
Good news, the code can work successfully, but there are some troubles for the results, the value of mdps are all 0 after the 60th sample point. Do you think is it also caused by the following part？ or other reasons.  Could you share with me your thoughts for how to debug next step? Thank you very much!

 if traffic_light_statuses(k)==1
          constraints=[constraints, 
                       implies(a{k}(1),[-5<=S1*x{k}<=0, mdps(k+1)==5]);];
          flag=1;
    elseif  traffic_light_statuses(k)==2||traffic_light_statuses(k)==3 
              flag=1;
          constraints=[constraints, 
                       implies(a{k}(2),[-5<=S1*x{k}<=0, mdps(k+1)==2]);];
     end
       constraints = [constraints, sum(a{k}) == 1];  
![image](https://github.com/yalmip/YALMIP/assets/148333150/e5538b84-e371-400d-bfd9-73663ea55114)

**Reply by johanlofberg** [reply]
I don't see why your logic would force mdps to be non-zero. if traffic_light_statuses(k)==1 then it can simply set a{k}(1) to 0 and mdps will be unconstrained

**Reply by SIYUANLI2023** [reply]
Hi Johan, good morning, I got the reason why the mdps are zero, as the slover cann't  find a solution. could I consult another question with you?

I rewrite those constrains, which are shown as follows. It seems like the first part conflict with the second part. When I keep both of them, the code can not work, but when I comment one of them, the code can work normally. I can not understand, could you please 
give me some guidance about it? Many thanks!
1.if traffic_light_statuses(k)==1
          constraints=[constraints, 
                       implies(-10<=S1*x{k}<=3, mdps(k+1)==5);];
          flag=1;
   elseif  traffic_light_statuses(k)==2||traffic_light_statuses(k)==3 
              flag=1;
          constraints=[constraints, 
                       implies(-10<=S1*x{k}<=3.6, mdps(k+1)==2);];
     end
 
 2.     constraints=[constraints, implies(4<=S1*x{k}<=7.2, mdps(k+1)==6);];

**Reply by johanlofberg** [reply]
Please put code in code blocks, it looks horrible now to read when unindented etc

I always recommend as flat models as possible. The second one I would write as (with d binary)

```
constraints=[constraints, implies(d(k,1), S1*x{k}<=4),
                          implies(d(k,2), [4 <= S1*x{k}<=7.2,mdps(k+1)==6] ),
                          implies(d(k,3), S1*x{k}>=7.2),
                          sum(d(k,:)) == 1]
```
https://yalmip.github.io/example/modellingif/

**Reply by SIYUANLI2023** [reply]
Thank you Johan, thanks for your patience, learn a lot from you.  
Best regards,
Lee.
